source/core/Native.cs

```diff
@@ -0,0 +1,711 @@
+            public static void Call(Hash hash, params InputArgument[] arguments)
+            {
+                Call<int>((ulong)hash, arguments);
+            }
+
+            internal static T Call<T>(ulong hash, params InputArgument[] arguments)
+            {
+                var task = new NativeTask();
+                task._hash = hash;
+                task._arguments = arguments;
+
+                ScriptDomain.CurrentDomain.ExecuteTask(task);
+
+                unsafe { return (T)(ObjectFromNative(typeof(T), task._result)); }
+            }
+
+            internal static ulong ObjectToNative(object value)
+            {
+                if (ReferenceEquals(value, null))
+                {
+                    return 0;
+                }
+
+                var type = value.GetType();
+
+                if (type.IsEnum)
+                {
+                    value = Convert.ChangeType(value, type = Enum.GetUnderlyingType(type));
+                }
+
+                if (type == typeof(bool))
+                {
+                    return (ulong)(((bool)value) ? 1 : 0);
+                }
+                if (type == typeof(int))
+                {
+                    return (ulong)((int)value);
+                }
+                if (type == typeof(uint))
+                {
+                    return (uint)value;
+                }
+                if (type == typeof(long))
+                {
+                    return (ulong)((long)value);
+                }
+                if (type == typeof(ulong))
+                {
+                    return (ulong)value;
+                }
+                if (type == typeof(float))
+                {
+                    return BitConverter.ToUInt32(BitConverter.GetBytes((float)value), 0);
+                }
+                if (type == typeof(Double))
+                {
+                    return BitConverter.ToUInt32(BitConverter.GetBytes((float)((double)value)), 0);
+                }
+
+                if (type == typeof(String))
+                {
+                    return (ulong)ScriptDomain.CurrentDomain.PinString((string)(value)).ToInt64();
+                }
+
+                if (type == typeof(IntPtr))
+                {
+                    return (ulong)((IntPtr)value).ToInt64();
+                }
+
+                if (typeof(INativeValue).IsAssignableFrom(type))
+                {
+                    return ((INativeValue)value).NativeValue;
+                }
+
+                throw new InvalidCastException(String.Concat("Unable to cast object of type '", type.FullName, "' to native value"));
+            }
+            internal static unsafe object ObjectFromNative(Type type, ulong* value)
+            {
+                if (type.IsEnum)
+                {
+                    type = Enum.GetUnderlyingType(type);
+                }
+
+                if (type == typeof(bool))
+                {
+                    unsafe
```

Comment:
No need for additional `unsafe` scopes if the whole method is `unsafe` already, right?